packages/completer/src/widget.ts

```diff
@@ -522,34 +522,31 @@ export class Completer extends Widget {
   private _cycle(direction: Private.scrollType): void {
     const items = this.node.querySelectorAll(`.${ITEM_CLASS}`);
     const index = this._activeIndex;
+    const last = items.length - 1;
     let active = this.node.querySelector(`.${ACTIVE_CLASS}`) as HTMLElement;
     active.classList.remove(ACTIVE_CLASS);
 
-    if (direction === 'up') {
-      this._activeIndex = index === 0 ? items.length - 1 : index - 1;
-    } else if (direction === 'down') {
-      this._activeIndex = index < items.length - 1 ? index + 1 : 0;
-    } else {
-      // Measure the number of items on a page.
-      const boxHeight = this.node.getBoundingClientRect().height;
-      const itemHeight = active.getBoundingClientRect().height;
-      const pageLength = Math.floor(boxHeight / itemHeight);
-
-      // Update the index
-      if (direction === 'pageUp') {
-        this._activeIndex = index - pageLength;
-      } else {
-        this._activeIndex = index + pageLength;
+    switch (direction) {
+      case 'up':
+        this._activeIndex = index === 0 ? last : index - 1;
+        break;
+      case 'down':
+        this._activeIndex = index < last ? index + 1 : 0;
+        break;
+      case 'pageUp':
+      case 'pageDown': {
+        // Measure the number of items on a page and clamp to the list length.
+        const container = this.node.getBoundingClientRect();
+        const current = active.getBoundingClientRect();
+        const page = Math.floor(container.height / current.height);
+        const sign = direction === 'pageUp' ? -1 : 1;
+        this._activeIndex = Math.min(Math.max(0, index + sign * page), last);
+        break;
       }
-      // Clamp to the length of the list.
-      this._activeIndex = Math.min(
-        Math.max(0, this._activeIndex),
-        items.length - 1
-      );
     }
 
     active = items[this._activeIndex] as HTMLElement;
-    active.classList.add(ACTIVE_CLASS);
+    active?.classList.add(ACTIVE_CLASS);
```

Comment:
Not sure here: is it be more useful for this to lead to an error in console so that users can report it? If `active` ever gets `null`/`undefined` it would indicate a regression in the logic, and would be more difficult to diagnose if this is silenced, right? Not worth halting this PR on this but just food for thought.

Reply:
I think this might not actually be necessary and might have been part of the debugging process of this PR. I'm playing around with it and if I can't generate an error, I'll remove the check. It never existed before.